File: 354-russian-doll-envelopes/354-russian-doll-envelopes.java
class Solution {
    public int BS(ArrayList<int[]> dp, int ele){
        int si=0;
        int ei=dp.size()-1;
        
        while(si<=ei){
            int mid=(si+ei)/2;
            
            if(dp.get(mid)[1]<ele){
                si=mid+1;
            } else {
                ei=mid-1;
            }
        }
        
        return si;
    }
    
    public int maxEnvelopes(int[][] envelopes) {
        int n=envelopes.length;
        
        Arrays.sort(envelopes,(int[] a, int [] b)->{
           if(a[0]==b[0]) return b[1]-a[1];
            return a[0]-b[0]; // width  
        });
        
        ArrayList<int[]> dp=new ArrayList<>();
        
        
        for(int i=0; i<n; i++){
            int nwidth=envelopes[i][0];
            int nheight=envelopes[i][1];
            
            int pos=BS(dp,nheight);
            // if(dp.size()>0)
            // System.out.println(dp.get(dp.size()-1)[0]+" "+nwidth);
            if(pos==dp.size() && (dp.size()==0 || (dp.get(dp.size()-1)[0]<nwidth))){
                dp.add(new int[]{nwidth,nheight});
            } else if(pos<dp.size()){
                dp.set(pos,new int[]{nwidth,nheight});
            }
        }
        
        return dp.size();
    }
    
}

File: 354-russian-doll-envelopes/354-russian-doll-envelopes.cpp
class Solution {
public:
    int BS(vector<int>& dp, int h){
    int l=0;
    int r=dp.size()-1;

    while(l<=r){
        int mid=(l+r)/2;

        if(dp[mid]<h){
            l=mid+1;
        } else {
            r=mid-1;
        }
    }

    return l;
}

int maxEnvelopes(vector<vector<int>>& envelopes) {

    sort(envelopes.begin(),envelopes.end(),[](vector<int>& a,vector<int>& b){
        if(a[0]==b[0]) return a[1] > b[1];

        return a[0] < b[0];
    }); 

    // Arrays.sort(envelopes,(int[] a, int[] b)->{
    //     if(a[0]==b[0]) return b[1] - a[1];

    //     return a[0] - b[0];
    // }); 

    int n=envelopes.size();  

    vector<int> dp;

    for(int i=0; i<n; i++){
        int h=envelopes[i][1];

        int idx=BS(dp,h);
        if(idx==dp.size()){
            dp.push_back(h);
        } else {
            dp[idx]=h;
        }
    }

    return dp.size();
}

};